cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -225,7 +232,8 @@ class ARROW_EXPORT FileSystem : public std::enable_shared_from_this<FileSystem>
   /// Create a directory and subdirectories.
   ///
   /// This function succeeds if the directory already exists.
-  virtual Status CreateDir(const std::string& path, bool recursive = true) = 0;
+  virtual Status CreateDir(const std::string& path, bool recursive) = 0;
+  Status CreateDir(const std::string& path) { return CreateDir(path, true); }
```

Comment:
Does this make any concrete difference? AFAIU, optional arguments don't change the callee's actual signature

Reply:
The style guide forbids default arguments on virtual functions. The main difference is a single point of truth for specifying default arguments, since otherwise it's possible to change the default argument in an override (usually an error) https://wandbox.org/permlink/0wqJJM2nu92uwAUA

Reply:
I had not noticed this from the style guide. Is it enforced in other places? Having to create multiple overloads is cumbersome...

Reply:
Well, there's a `clang-tidy` pass which warns for this.

Reply:
I suppose it's not enabled in our lint tests then :-)

Reply:
I think clang-tidy issues so many warnings that most ignore them :D

Reply:
All power to you for changing this @bkietz. Default arguments in virtual functions are so confusing!